Subject: General platform-related issues
Author: Steve Greenland <stevegr@debian.org>
Last-Update: 2010-04-11

Various platform-related fixes, additions or updates. These are quite numerous
because the upstream code is from 1993.

Bug-Debian: http://bugs.debian.org/64382
Bug-Debian: http://bugs.debian.org/50240

Acked-by: Christian Kastner <debian@kvr.at>
Index: patched/externs.h
===================================================================
--- patched.orig/externs.h	2010-05-06 18:18:16.268435775 +0200
+++ patched/externs.h	2010-05-06 18:18:22.672435408 +0200
@@ -20,6 +20,7 @@
 # include <unistd.h>
 # include <string.h>
 # include <dirent.h>
+# include <errno.h>
 # define DIR_T	struct dirent
 # define WAIT_T	int
 # define WAIT_IS_INT 1
@@ -55,6 +56,7 @@
 extern	void		perror(), exit(), free();
 extern	char		*getenv(), *strcpy(), *strchr(), *strtok();
 extern	void		*malloc(), *realloc();
+
 # define SIG_T	void
 # define TIME_T	long
 # define PID_T int
Index: patched/cron.c
===================================================================
--- patched.orig/cron.c	2010-05-06 18:18:16.252435886 +0200
+++ patched/cron.c	2010-05-06 18:18:22.672435408 +0200
@@ -24,7 +24,7 @@
 
 
 #include "cron.h"
-#include <sys/signal.h>
+#include <signal.h>
 #if SYS_TIME_H
 # include <sys/time.h>
 #else
Index: patched/popen.c
===================================================================
--- patched.orig/popen.c	2010-05-06 18:18:16.279435871 +0200
+++ patched/popen.c	2010-05-06 18:18:22.672435408 +0200
@@ -29,7 +29,11 @@
 #endif /* not lint */
 
 #include "cron.h"
-#include <sys/signal.h>
+#include <signal.h>
+
+#if defined(BSD) || defined(POSIX)
+#  include <grp.h>
+#endif
 
 
 #define WANT_GLOBBING 0
Index: patched/do_command.c
===================================================================
--- patched.orig/do_command.c	2010-05-06 18:18:16.263436081 +0200
+++ patched/do_command.c	2010-05-06 18:18:22.672435408 +0200
@@ -21,7 +21,8 @@
 
 
 #include "cron.h"
-#include <sys/signal.h>
+#include <signal.h>
+#include <grp.h>
 #if defined(sequent)
 # include <sys/universe.h>
 #endif
@@ -207,7 +208,7 @@
 		 * we set uid, we've lost root privledges.
 		 */
 		setgid(e->gid);
-# if defined(BSD)
+# if defined(BSD) || defined(POSIX)
 		initgroups(env_get("LOGNAME", e->envp), e->gid);
 # endif
 		setuid(e->uid);		/* we aren't root after this... */
Index: patched/database.c
===================================================================
--- patched.orig/database.c	2010-05-06 18:18:16.257436063 +0200
+++ patched/database.c	2010-05-06 18:18:22.672435408 +0200
@@ -28,9 +28,20 @@
 #include <sys/stat.h>
 #include <sys/file.h>
 
-
 #define TMAX(a,b) ((a)>(b)?(a):(b))
 
+/* Try to get maximum path name -- this isn't really correct, but we're
+going to be lazy */
+
+#ifndef PATH_MAX
+
+#ifdef MAXPATHLEN
+#define PATH_MAX MAXPATHLEN 
+#else
+#define PATH_MAX 2048
+#endif
+
+#endif /* ifndef PATH_MAX */
 
 static	void		process_crontab __P((char *, char *, char *,
 					     struct stat *,
@@ -102,7 +113,7 @@
 
 	while (NULL != (dp = readdir(dir))) {
 		char	fname[MAXNAMLEN+1],
-			tabname[MAXNAMLEN+1];
+			tabname[PATH_MAX+1];
 
 		/* avoid file names beginning with ".".  this is good
 		 * because we would otherwise waste two guaranteed calls
Index: patched/compat.h
===================================================================
--- patched.orig/compat.h	2010-05-06 18:18:16.247436070 +0200
+++ patched/compat.h	2010-05-06 18:18:22.672435408 +0200
@@ -62,8 +62,8 @@
 #endif
 
 #ifndef POSIX
-# if (BSD >= 199103) || defined(__linux) || defined(ultrix) || defined(AIX) ||\
-	defined(HPUX) || defined(CONVEX) || defined(IRIX)
+# if (BSD >= 199103) || defined(__linux__) || defined(__GNU__) || defined(ultrix) ||\
+        defined(AIX) ||\ defined(HPUX) || defined(CONVEX) || defined(IRIX) || defined(__GLIBC__)
 #  define POSIX
 # endif
 #endif
@@ -76,17 +76,17 @@
 
 /*****************************************************************/
 
-#if !defined(BSD) && !defined(HPUX) && !defined(CONVEX) && !defined(__linux)
+#if !defined(BSD) && !defined(HPUX) && !defined(CONVEX) && !defined(__linux__) && !defined(__GNU__)
 # define NEED_VFORK
 #endif
 
-#if (!defined(BSD) || (BSD < 198902)) && !defined(__linux) && \
-	!defined(IRIX) && !defined(NeXT) && !defined(HPUX)
+#if (!defined(BSD) || (BSD < 198902)) && !defined(__linux__) && \
+	!defined(IRIX) && !defined(NeXT) && !defined(HPUX) && !defined(__GNU__) && !defined(__GLIBC__)
 # define NEED_STRCASECMP
 #endif
 
-#if (!defined(BSD) || (BSD < 198911)) && !defined(__linux) &&\
-	!defined(IRIX) && !defined(UNICOS) && !defined(HPUX)
+#if (!defined(BSD) || (BSD < 198911)) && !defined(__linux__) &&\
+	!defined(IRIX) && !defined(UNICOS) && !defined(HPUX) && !defined(__GNU__) && !defined(__GLIBC__)
 # define NEED_STRDUP
 #endif
 
@@ -102,19 +102,19 @@
 # define NEED_SETSID
 #endif
 
-#if (defined(POSIX) && !defined(BSD)) && !defined(__linux)
+#if (defined(POSIX) && !defined(BSD)) && !defined(__linux__) && !defined(__GNU__) && !defined(__GLIBC__)
 # define NEED_GETDTABLESIZE
 #endif
 
-#if (BSD >= 199103)
+#if (BSD >= 199103) || defined(__linux)
 # define HAVE_SAVED_UIDS
 #endif
 
-#if !defined(ATT) && !defined(__linux) && !defined(IRIX) && !defined(UNICOS)
+#if !defined(ATT) && !defined(__linux__) && !defined(__GNU__) && !defined(IRIX) && !defined(UNICOS) && !defined(__GLIBC__)
 # define USE_SIGCHLD
 #endif
 
-#if !defined(AIX) && !defined(UNICOS)
+#if !defined(AIX) && !defined(UNICOS) && !defined(DEBIAN)
 # define SYS_TIME_H 1
 #else
 # define SYS_TIME_H 0
Index: patched/pathnames.h
===================================================================
--- patched.orig/pathnames.h	2010-05-06 18:18:16.273435823 +0200
+++ patched/pathnames.h	2010-05-06 18:18:22.673435429 +0200
@@ -19,7 +19,7 @@
  * $Id: pathnames.h,v 1.3 1994/01/15 20:43:43 vixie Exp $
  */
 
-#if (defined(BSD)) && (BSD >= 199103) || defined(__linux) || defined(AIX)
+#if (defined(BSD)) && (BSD >= 199103) || defined(__linux__) || defined(AIX) || defined(__GNU__) || defined(__GLIBC__)
 # include <paths.h>
 #endif /*BSD*/
 
